Caching (#246)
avantifellows/plio-backend: docs/CACHING.md

## Caching
Plio uses the Redis cache backend powered by the [Django Redis](https://github.com/jazzband/django-redis) package.

This guide aims to provide details on how Plio is using caching and pre-requisites for someone contributing to the code.

### Caching workflow
Plio's caching mechanism can be explained in three simple steps:
1. Receives an API request
2. Searches for the requested data in the cache
3. If not present, query the database and save the response in cache for future requests

> ![Overview of caching](images/caching-workflow.png)

Reviewer: One thing is not clear:
- what happens if the value is present in the cache but has been updated since it was put in the cache?

Author: Okay, makes sense. I'll update it or maybe create a separate one to explain the flow better.

> 

Reviewer: can you make a list of requests that we are caching? 

GET plio
GET user

### Cache keys
The calculation of the cache keys are based on the model instances. For example, an instance for plio ID: 1 will have `plio_1` as the cache key.
For more details, check out the `get_cache_key` function in `plio/cache.py`.

### Cache refresh or invalidation

> When a particular instance is updated, it's cached value gets deleted along with any other related instances cache that depends on this cache. For example, if a session instance cache uses plio instance cache in it's response, if the plio is modified, both plio instance and session instance cache will be deleted.

Reviewer: Just wording it a bit more clearly.

```suggestion
When a particular instance is updated, its cached value gets deleted along with the cache of any other related instances that depends on this cache. For example, consider a session instance cache that uses a plio instance cache. Now, if the plio is modified, the caches for both the plio instance and the session instance will be deleted.
```

> The new cache will be set during the first fresh response calculated from the database and will be used for subsequent requests.

Reviewer: ```suggestion
The new cache will be set when the first fresh response is calculated from the database and will be used for subsequent requests.
```

2. Searches for the requested data in the cache
3. If not present, query the database and save the response in cache for future requests

##### Caching workflow explained via flow diagram

> ![Overview of caching](images/caching-workflow.png)

Reviewer: Here, the part about cache invalidation is still missing in the diagram.

For more details, check out the `get_cache_key` function in `plio/cache.py`.

### Cache invalidation
When a particular instance is updated, its cached value gets deleted along with the cache of any other related instances that depends on this cache. For example, consider a session instance cache that uses a plio instance cache. Now, if the plio is modified, the caches for both the plio instance and the session instance will be deleted.

The new cache will be set when the first fresh response is calculated from the database and will be used for subsequent requests.

##### Cache invalidation explained via flow diagram
![Overview of caching](images/cache-invalidation-workflow.png)

### Current cached data

> We only have implemented caching for data that has high GET requests. The following resources have been cached:

Reviewer: ```suggestion
We have only implemented caching for models with a high number of GET requests. The following resources have been cached:
```

### Current cached data
We have only implemented caching for models with a high number of GET requests. The following resources have been cached:
1. #### Plio
    - Plio cache is created when there is a retrieve request for a plio instance.
    - Plio cache is re-created when there is a create, update or delete request for a plio instance.
    - Plio cache is deleted when there is a create, update or delete request for any of the following related instances for a plio instance:
        - Video
        - Item
        - Question
        - User

2. #### User
    - User cache is created when there is a retrieve request for a user instance.
    - User cache is re-created when there is a create, update or delete request for any of the following related instances:
      - User

>       - OrganizationUser (as the UserSerializer is called when modifying OrganizationUser)

Reviewer: ```suggestion
      - OrganizationUser (as the `UserSerializer` is called when `OrganizationUser` is modified)
```
